collect_app/src/main/java/org/odk/collect/android/utilities/FileUtils.java

```diff
@@ -466,4 +467,101 @@ public static String getFilenameError(String filename) {
         boolean containsPossiblyRestricted = CharMatcher.anyOf(possiblyRestricted).matchesAnyOf(possiblyRestricted);
         return "Problem with project name file. Contains @: " + containsAt + ", Contains non-ascii: " + containsNonAscii + ", Contains restricted: " + containsPossiblyRestricted;
     }
+
+    /**
+     * Common routine to take a downloaded document save the contents in the file
+     * 'file'.
+     * <p>
+     * The file is saved into a temp folder and is moved to the final place if everything
+     * is okay, so that garbage is not left over on cancel.
+     */
+    public static void interuptablyWriteFile(InputStream inputStream, File destinationFile, File tempDir, OngoingWorkListener listener)
```

Comment:
I'm still not sure if we actually need to be this cautious - I think we could just clean up after ourselves on a cancel, but I'm still not feeling confident enough to rip the extra temp dir out.